libraries/AP_PacketDigitalCAN/AP_PacketDigitalCAN.cpp

```diff
@@ -0,0 +1,609 @@
+        txFrame.data[3] = LOWBYTE(value);
+        }
+        break;
+    }
+
+    if (serialnumber == 0) {
+        // send to all
+        txFrame.id = 0x00240000;
+    } else {
+        txFrame.id = 0x00210000 | (uint32_t)serialnumber;
+    }
+
+    txFrame.id |= AP_HAL::CANFrame::FlagEFF;
+
+    txFrame.data[0] = ++sequence_prev;
+    txFrame.data[1] = (uint8_t)type;
+    txFrame.dlc += 2;
+
+    //gcs().send_text(MAV_SEVERITY_INFO, "PDCAN: %u Tx Seq %u 0x%2X", (unsigned)serialnumber, (unsigned)txFrame.data[0], (unsigned)txFrame.data[1]);
+
+    if (write_frame(txFrame, 50000)) {
+        // keep track of what we sent last in case we get an ACK/NACK
+        packet_sent_prev = type;
+    } else {
+        //gcs().send_text(MAV_SEVERITY_DEBUG, "PDCAN: %u Tx Seq %u 0x%2X failed", (unsigned)serialnumber, (unsigned)txFrame.data[0], (unsigned)txFrame.data[1]);
+    }
+}
+
+// get MPPT_device index by serial number.
+// if serial number found in MPP_devices list, return the 0 indexed value
+// else return UINT8_MAX
+uint8_t AP_PacketDigitalCAN::get_device_index(const uint16_t serial_number) const
+{
+    for (uint8_t i=0; i<device_count; i++) {
+        if (MPPT_devices[i].serialnumber == serial_number) {
+            return i;
+        }
+    }
+    return UINT8_MAX;
+}
+
+// write frame on CAN bus, returns true on success
+bool AP_PacketDigitalCAN::write_frame(AP_HAL::CANFrame &out_frame, const uint64_t timeout)
+{
+    if (!_initialized) {
+        debug_can(AP_CANManager::LOG_ERROR, "Driver not initialized for write_frame\n\r");
+        return false;
+    }
+
+    bool read_select = false;
+    bool write_select = true;
+    bool ret;
+    do {
+        ret = _can_iface->select(read_select, write_select, &out_frame, AP_HAL::native_micros64() + timeout);
+        if (!ret || !write_select) {
+            hal.scheduler->delay_microseconds(200);
+        }
+    } while (!ret || !write_select);
+
+    uint64_t deadline = AP_HAL::native_micros64() + 2000000;
+    return (_can_iface->send(out_frame, deadline, AP_HAL::CANIface::AbortOnError) == 1);
+}
+
+// read frame on CAN bus, returns true on succses
+bool AP_PacketDigitalCAN::read_frame(AP_HAL::CANFrame &recv_frame, const uint64_t timeout, AP_HAL::CANIface::CanIOFlags &flags)
+{
+    if (!_initialized) {
+        debug_can(AP_CANManager::LOG_ERROR, "Driver not initialized for read_frame\n\r");
+        return false;
+    }
+    // wait for space in buffer to read
+    bool read_select = true;
+    bool write_select = false;
+    bool ret = _can_iface->select(read_select, write_select, nullptr, AP_HAL::native_micros64() + timeout);
+    if (!ret || !read_select) {
+        // return false if no data is available to read
+        return false;
+    }
+
+    uint64_t time;
+
+    // read frame and return success
+    return (_can_iface->receive(recv_frame, time, flags) == 1);
+}
+
+bool AP_PacketDigitalCAN::pre_arm_check(char* reason, uint8_t reason_len) const
```

Comment:
You're not filling in the reason here.  (which can be bad if reason isn't zero-initialised :-) )

Reply:
good catch. fixed.